File: src/main/java/icu/suc/servergui/ClickType.java
package icu.suc.servergui;

import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public enum ClickType {
    LEFT,
    SHIFT_LEFT,
    RIGHT,
    SHIFT_RIGHT,
    MIDDLE,
    DOUBLE,
    THROW,
    THROW_ALL,
    SWAP;

    public static @Nullable ClickType of(net.minecraft.world.inventory.@NotNull ClickType type, byte button) {
        switch (type) {
            case PICKUP -> {
                switch (button) {
                    case 0 -> {
                        return LEFT;
                    }
                    case 1 -> {
                        return RIGHT;
                    }
                }
            }
            case QUICK_MOVE -> {
                switch (button) {
                    case 0 -> {
                        return SHIFT_LEFT;
                    }
                    case 1 -> {
                        return SHIFT_RIGHT;
                    }
                }
            }
            case SWAP -> {
                return SWAP;
            }
            case CLONE -> {
                return MIDDLE;
            }
            case THROW -> {
                switch (button) {
                    case 0 -> {
                        return THROW;
                    }
                    case 1 -> {
                        return THROW_ALL;
                    }
                }
            }
            case PICKUP_ALL -> {
                if (button == 0) {
                    return DOUBLE;
                }
            }
        }

        return null;
    }
}

File: src/main/java/icu/suc/servergui/ClickType.kt
package icu.suc.servergui

import net.minecraft.world.inventory.ClickType.*

enum class ClickType {
    LEFT,
    SHIFT_LEFT,
    RIGHT,
    SHIFT_RIGHT,
    MIDDLE,
    DOUBLE,
    THROW,
    THROW_ALL,
    SWAP;

    companion object {
        fun of(type: net.minecraft.world.inventory.ClickType, button: Int): ClickType? = when (type) {
            PICKUP -> when (button) {
                0 -> LEFT
                1 -> RIGHT
                else -> null
            }
            QUICK_MOVE -> when (button) {
                0 -> SHIFT_LEFT
                1 -> SHIFT_RIGHT
                else -> null
            }
            net.minecraft.world.inventory.ClickType.SWAP -> SWAP
            CLONE -> MIDDLE
            net.minecraft.world.inventory.ClickType.THROW -> when (button) {
                0 -> THROW
                1 -> THROW_ALL
                else -> null
            }
            PICKUP_ALL ->
                if (button == 0) DOUBLE else null

            QUICK_CRAFT -> null
        }
    }
}